README.rst
========================
asyncio-fast-portscanner
========================


.. image:: https://img.shields.io/pypi/v/asyncio-fast-portscanner.svg
        :target: https://pypi.python.org/pypi/asyncio-fast-portscanner

.. image:: https://img.shields.io/travis/Razikus/asyncio-fast-portscanner.svg
        :target: https://travis-ci.org/Razikus/asyncio-fast-portscanner

.. image:: https://readthedocs.org/projects/asyncio-fast-portscanner/badge/?version=latest
        :target: https://asyncio-fast-portscanner.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




Get your active host list in N * timeout fixed time


* Free software: MIT license
* Documentation: https://asyncio-fast-portscanner.readthedocs.io.


Features
--------

* Scan specific ports for specific hosts with fixed timeout






HISTORY.rst
=======
History
=======

0.0.2 (2019-10-19)
------------------

* Fixed cli version.

0.0.1 (2019-10-19)
------------------

* First release on PyPI.


tests/__init__.py
# -*- coding: utf-8 -*-

"""Unit test package for asyncio_fast_portscanner."""


tests/test_asyncio_fast_portscanner.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `asyncio_fast_portscanner` package."""

import pytest

from click.testing import CliRunner

from asyncio_fast_portscanner import asyncio_fast_portscanner
from asyncio_fast_portscanner import cli
import json


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    """ Test for RANGE + additional port"""
    result = runner.invoke(cli.scan, ["192.168.0-2.0-255", "22", "-r", "RANGE", "-t", "2", "-a"])
    assert result.exit_code == 0
    unJsoned = json.loads(result.output)
    assert len(unJsoned) == 256 * 3

    """ Test for smaller range"""
    result = runner.invoke(cli.scan, ["192.168.2.1-128", "22", "-r", "RANGE", "-t", "0.3", "-a"])
    assert result.exit_code == 0
    unJsoned = json.loads(result.output)
    assert len(unJsoned) == 128

    """ Test for CIDR + additional port"""
    result = runner.invoke(cli.scan, ["192.168.0.0/16", "22", "33", "-r", "CIDR", "-t", "5", "-a"])
    assert result.exit_code == 0
    unJsoned = json.loads(result.output)
    assert len(unJsoned) == 256 * 256


asyncio_fast_portscanner/asyncio_fast_portscanner.py
# -*- coding: utf-8 -*-
import asyncio
import ipaddress
from typing import Sequence, Tuple
import itertools
import re

"""Main module."""


class FastPortScanner():
    def __init__(self, timeout: float, verboseLogging: bool = False, hostList: Sequence[str] = [],
                 portList: Sequence[int] = []):
        self.hostList = hostList
        self.portList = portList
        self.timeout = timeout
        self.verboseLogging = verboseLogging

    async def loadHostListByCidr(self, cidr: str) -> Tuple[bool, str]:
        self.hostList = []
        try:
            self.hostList = [str(ip) for ip in ipaddress.IPv4Network(cidr)]
            if (self.verboseLogging):
                print("Hostlist:", self.hostList)
        except ipaddress.AddressValueError as e:
            return False, e
        except ipaddress.NetmaskValueError as e:
            return False, e
        except ValueError as e:
            return False, e
        if (len(self.hostList) <= 0):
            return False, "List of hosts is empty"

        return True, ""

    async def loadPortList(self, ports: Sequence[int]) -> Tuple[bool, str]:
        self.portList = []
        if (len(ports) == 0):
            return False, "0 length sequence"
        self.portList = ports
        if (self.verboseLogging):
            print("Hostlist:", self.portList)
        return True, ""

    # This shit is to refactor, cause i can't invent anything better now :D


    async def loadHostListByRange(self, rangeString: str) -> Tuple[bool, str]:
        splitted = rangeString.split(".")
        hostList = []
        indexes = []
        lists = []
        index = 0
        for piece in splitted:
            if ("-" in piece):
                pieceSplitted = piece.split("-")
                minimum = int(pieceSplitted[0])
                maximum = int(pieceSplitted[1])
                if(minimum > maximum):
                    return False, "One of range is invalid (minimum > maximum)"

                lists.append([x for x in range(minimum, maximum + 1)])
                indexes.append(index)
            index = index + 1

        pipe = "-0-.-1-.-2-.-3-"
        for index in range(0, 4):
            if index not in indexes:
                pipe = pipe.replace("-" + str(index) + "-", splitted[index])

        if (len(indexes) > 0):
            products = itertools.product(*lists)
            for product in products:
                pipeFormat = pipe
                indexNow = 0
                for index in indexes:
                    pipeFormat = pipeFormat.replace("-" + str(index) + "-", str(product[indexNow]))
                    indexNow = indexNow + 1
                hostList.append(pipeFormat)
        else:
            hostList.append(pipe)

        ipRegex = re.compile("^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$")
        for host in hostList:
            if(not ipRegex.match(host)):
                return False, host + " doesn't match ipv4 regex"

        if len(hostList) <= 0:
            return False, "List of hosts is empty"

        self.hostList = hostList
        return True, ""


    async def tcp_check(hostname: str, port: int) -> bool:
        try:
            reader, writer = await asyncio.open_connection(hostname, port)
            writer.close()
            return True
        except Exception as e:
            return False
        return True


    async def tcp_check_timeout(host: str, port: int, timeout: float) -> dict:
        try:
            return {"host": host,
                    "status": await asyncio.wait_for(FastPortScanner.tcp_check(host, port), timeout=timeout),
                    "port": port}
        except asyncio.TimeoutError:
            return {"host": host, "status": False, "port": port}


    async def gatherResults(self):
        tasks = []
        for host in self.hostList:
            for port in self.portList:
                if (self.verboseLogging):
                    print("Preparing task for " + host + ":" + str(port) + "...")
                tasks.append(FastPortScanner.tcp_check_timeout(host, port, self.timeout))
        return await asyncio.gather(*tasks)


asyncio_fast_portscanner/cli.py
# -*- coding: utf-8 -*-

"""Console script for asyncio_fast_portscanner."""
import sys
import click
from .asyncio_fast_portscanner import FastPortScanner
import asyncio
import json
import itertools

@click.group()
def main():
    pass


@main.command()
@click.argument("range", type=click.STRING)
@click.argument("ports", nargs=-1, metavar="ports", type=click.IntRange(min=0, max=65535))
@click.option("--timeout", "-t", default=0.3, help="Specifies timeout", type=click.FLOAT)
@click.option("--outputformat", "-f", default="JSON",
              help="Specifies output format. \nFor json: \n{'host': host, 'ports': { 22: true, 33: false}}",
              type=click.Choice(["JSON", "TEXT"], case_sensitive=False))
@click.option("--sockettype", "-s", default="TCP", help="Specifies TCP or UDP type",
              type=click.Choice(["TCP", "UDP"], case_sensitive=False))
@click.option("--rangetype", "-r", default="cidr", help="Specifies type of range",
              type=click.Choice(["CIDR", "RANGE"], case_sensitive=False))
@click.option("--verbose", "-v", default=False, is_flag=True)
@click.option("--activeOnly", "-a", default=True, is_flag=True, help="Returns if at least one of the port is active, default true")
def scan(range: str, ports, timeout, outputformat: str, sockettype: str, rangetype: str, verbose: bool, activeonly: bool):
    """Scans for hosts in specified range.

    Range in CIDR format like 192.168.1.1/24
    So for example 192.168.1.0/24 gives 192.168.1.0 - 192.168.1.255 results

    Range in RANGE format like 192.168.1.0-255
    So for example 192.168.1.0-255 gives 192.168.1.0 - 192.168.1.255 results

    Range in RANGE format like 192.168.1-2.0-255
    This example produces 192.168.1.0 - 192.168.1.255 and 192.168.2.0 - 192.168.2.255 results

    For bigger bunch of tasks you have to increase timeout, because it will always report as Closed (because task
    will never have a chance to execute = timeout)

    For example for scan 192.168.0-255.0-255 22 -r RANGE -v -t 5 - timeout of 10 seconds is enough

    """
    loop = asyncio.get_event_loop()
    scanner = FastPortScanner(timeout, verbose)
    if rangetype.upper() == "CIDR":
        result = loop.run_until_complete(scanner.loadHostListByCidr(range))
        if not result[0]:
            conditionalClickEcho(verbose, result[1])
            sys.exit(1)
    elif rangetype.upper() == "RANGE":
        result = loop.run_until_complete(scanner.loadHostListByRange(range))
        if not result[0]:
            conditionalClickEcho(verbose, result[1])
            sys.exit(1)

    result = loop.run_until_complete(scanner.loadPortList(ports))
    if not result[0]:
        conditionalClickEcho(verbose, result[1])
        sys.exit(1)

    results = loop.run_until_complete(scanner.gatherResults())
    grouped = groupResults(results, activeonly)
    if(outputformat == "JSON"):
        click.echo(json.dumps(grouped))
    elif(outputformat == "TEXT"):
        for item in grouped:
            print(item, end = "; ")
            for port in grouped[item]["ports"]:
                if(grouped[item]["ports"][port]):
                    print(str(port), "Open", end = "; ")
                else:
                    print(str(port), "Closed", end = "; ")
            print()




def conditionalClickEcho(verbose, message):
    if verbose:
        click.echo(message)




def groupResults(result, activeOnly):
    grouped = dict()
    for obj in result:
        host = obj["host"]
        port = obj["port"]
        status = obj["status"]
        if host in grouped:
            grouped[host]["ports"][port] = status
        else:
            grouped[host] = dict()
            grouped[host]["ports"] = {port: status}

    if(activeOnly):
        toRemove = [key for key in grouped if not checkForActive(grouped[key]["ports"])]
        for key in toRemove: del grouped[key]

    return grouped

def checkForActive(portsStatus):
    for key in portsStatus:
        if(portsStatus[key]):
            return True
    return False


if __name__ == "__main__":
    main()  # pragma: no cover
